src/gui/SampleLoader.cpp

```diff
@@ -0,0 +1,136 @@
+#include "FileDialog.h"
+#include "PathUtil.h"
+#include "Song.h"
+
+namespace lmms::gui {
+QString SampleLoader::openAudioFile(const QString& previousFile)
+{
+	auto openFileDialog = gui::FileDialog(nullptr, QObject::tr("Open audio file"));
+	auto dir = QString{};
+
+	if (!previousFile.isEmpty())
+	{
+		auto file = QString{previousFile};
+		if (QFileInfo{file}.isRelative())
+		{
+			file = ConfigManager::inst()->userSamplesDir() + file;
+			if (!QFileInfo{file}.exists()) { file = ConfigManager::inst()->factorySamplesDir() + previousFile; }
+		}
+
+		dir = QFileInfo{file}.absolutePath();
+	}
+	else { dir = ConfigManager::inst()->userSamplesDir(); }
+
+	// change dir to position of previously opened file
+	openFileDialog.setDirectory(dir);
+	openFileDialog.setFileMode(gui::FileDialog::ExistingFiles);
+
+	// set filters
+	// TODO: Since libsndfile 1.1.0, MP3 is supported
+	const auto fileTypes = QStringList{QObject::tr("All Audio-Files (*.wav *.ogg *.ds *.flac *.spx *.voc "
+												   "*.aif *.aiff *.au *.raw)"),
+		QObject::tr("Wave-Files (*.wav)"), QObject::tr("OGG-Files (*.ogg)"), QObject::tr("DrumSynth-Files (*.ds)"),
+		QObject::tr("FLAC-Files (*.flac)"), QObject::tr("SPEEX-Files (*.spx)"), QObject::tr("VOC-Files (*.voc)"),
+		QObject::tr("AIFF-Files (*.aif *.aiff)"), QObject::tr("AU-Files (*.au)"), QObject::tr("RAW-Files (*.raw)")};
+
+	openFileDialog.setNameFilters(fileTypes);
+
+	if (!previousFile.isEmpty())
+	{
+		// select previously opened file
+		openFileDialog.selectFile(QFileInfo{previousFile}.fileName());
+	}
+
+	if (openFileDialog.exec() == QDialog::Accepted)
+	{
+		if (openFileDialog.selectedFiles().isEmpty()) { return ""; }
+
+		return PathUtil::toShortestRelative(openFileDialog.selectedFiles()[0]);
+	}
+
+	return "";
+}
+
+QString SampleLoader::openWaveformFile(const QString& previousFile)
+{
+	return openAudioFile(
+		previousFile.isEmpty() ? ConfigManager::inst()->factorySamplesDir() + "waveforms/10saw.flac" : previousFile);
+}
+
+std::unique_ptr<SampleBuffer> SampleLoader::createBufferFromFile(const QString& filePath, bool collectErrorWhenNotFound)
+{
+	if (filePath.isEmpty()) { return std::make_unique<SampleBuffer>(); }
+
+	if (collectErrorWhenNotFound)
+	{
+		auto absolutePath = PathUtil::toAbsolute(filePath);
+		if (!QFileInfo(absolutePath).exists())
+		{
+			QString message = QObject::tr("Sample not found: %1").arg(filePath);
+			Engine::getSong()->collectError(message);
+			return std::make_unique<SampleBuffer>();
+		}
+	}
+
+	try
+	{
+		return std::make_unique<SampleBuffer>(filePath);
+	}
+	catch (const std::runtime_error& error)
+	{
+		displayError(QString::fromStdString(error.what()));
+		return std::make_unique<SampleBuffer>();
+	}
+}
+
+std::unique_ptr<SampleBuffer> SampleLoader::createBufferFromBase64(const QString& base64, int sampleRate)
+{
+	if (base64.isEmpty()) { return std::make_unique<SampleBuffer>(); }
+
+	try
+	{
+		return std::make_unique<SampleBuffer>(base64.toUtf8().toBase64(), sampleRate);
```

Comment:
Isn't this supposed to be `QByteArray::fromBase64(base64.toUtf8())` to match what the constructor does?

Reply:
Good catch. Thank you. `SampleBuffer::SampleBuffer(const QByteArray&, int)` does expect the data to be already decoded. But in this function, the data we return is in Base64 form and not the decoded sample data. I think I can simplify this though: I made a separate constructor that took raw Base64 data originally because it was conflicting with the audio file constructor, but since we should pass in a sample rate for the Base64 constructor, the two constructors can take in a `QString` yet still be differentiated.